lib/geo_postgis/geometry.ex

```diff
@@ -78,15 +78,31 @@ if Code.ensure_loaded?(Ecto.Type) do
     def cast(%struct{} = geom) when struct in @geometries, do: {:ok, geom}
 
     def cast(%{"type" => type, "coordinates" => _} = geom) when type in @types do
-      {:ok, Geo.JSON.decode!(geom)}
+      do_cast(geom)
     end
 
     def cast(%{"type" => "GeometryCollection", "geometries" => _} = geom) do
-      {:ok, Geo.JSON.decode!(geom)}
+      do_cast(geom)
     end
 
     def cast(geom) when is_binary(geom) do
-      {:ok, geom |> Geo.PostGIS.Config.json_library().decode!() |> Geo.JSON.decode!()}
+      do_cast(geom)
+    end
+
+    defp do_cast(geom) when is_binary(geom) do
+      with {:ok, geom} <- Geo.PostGIS.Config.json_library().decode(geom),
+           {:ok, result} <- Geo.JSON.decode(geom) do
+        {:ok, result}
+      else
+        _ -> :error
+      end
```

Comment:
Probably a good idea for this to delegate to the catch-all function head below, just in case we ever want to give finer-grained errors (returning a keyword list) on failures:

```suggestion
      case Geo.PostGIS.Config.json_library().decode(geom) do
        {:ok, geom} when is_map(geom) -> do_cast(geom)
        _ -> :error
      end
```

Reply:
What about this? That way we also have details of an error while decoding json. Although we can theoretically return `{:ok, not_map}`.
```suggestion
      case Geo.PostGIS.Config.json_library().decode(geom) do
        {:ok, geom} when is_map(geom) -> do_cast(geom)
        error -> error
      end
```

Reply:
Mmm, I like that. [the docs for `Ecto.Type.cast/1`](https://hexdocs.pm/ecto/3.11.2/Ecto.Type.html#c:cast/1) indicate we need to return a keyword list of errors, optionally including a `:message`, so I think it'd need to be something like:

```suggestion
      case Geo.PostGIS.Config.json_library().decode(geom) do
        {:ok, geom} when is_map(geom) -> do_cast(geom)
        {:error, reason} -> [message: "failed to decode JSON", reason: reason]
      end
```

Reply:
Ah, right you are!